### PR TITLE
Allow regex in blacklisted dirs and files

### PR DESCRIPTION
**Pull request by tdaget**
## Description
I need to support regex in blacklist functionality for files and dirs. Need to set blacklist on the server (want to enforce corporate IT policies).
`excluded_directories` and `blacklisted_files` variables already exist in the parameter file but only work on exact folder name and filename.

In lib/private/Files/Filesystem.php : envolve `isFileBlacklisted` function in order to evaluate regex match for each entry declared in $excluded_directories and $blacklisted_files config parameters respectively for folder name and filename.
In config/config/sample.php : For $excluded_directories and $blacklisted_files variables, add chars to be regex compatible.

Regex syntax can be tested with [https://www.phpliveregex.com](https://www.phpliveregex.com)

## Motivation and Context
This change was initiated as in our structure several people sync there local Outlook pst file (despite IT recommendations). This generate a lot (a very lot) of file's change on local disk so made a lot of network traffic between Owncloud desktop client and server.

## How Has This Been Tested?
Currently, this patch is deployed on our test platform without any drawbacks. Community comments are welcome ...
Note: Regex match is CPU consuming, hope this will not raise major problem on a production platform. Stress tests must be performed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: lib/private/Files/Filesystem.php [hunk 1]
```diff
@@ -655,20 +655,26 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
...
-		$blacklist = \array_map('trim', $blacklist);
-		$blacklist = \array_map('strtolower', $blacklist);
-		$match = \array_intersect($path_parts, $blacklist);
-		if ($match) {
-			return true;
-		}
+                if ($excluded) {
+                        $excluded = \array_map('trim', $excluded);
+                        $excluded = \array_map('strtolower', $excluded);
+                        foreach($path_parts as $path_part) {
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: * You could do a check here (and below) if `$excluded` contains a to be defined "forbidden" character (out of a list, for hints see: https://github.com/owncloud/core/issues/29759 plus usual regex characters, blacklist or whitelist what ever better fits) which would indicate a present regex and then branch between the old and your new regex querey. This would greatly improve performance as it continues using the old way with `array_intersect` and uses `preg_match` in the loop only if necessary.
* The query characters used for indicating a regex must be described in the config.sample file above.

* Add an example like:
```
hugo   // does not contain a regex character
will match the first item of the path component only:
hugo
hugo.test
testhugo

(?:hugo)  // contains regex
will match all items of the path component:
hugo
hugo.test
testhugo
```

File: config/config.sample.php [hunk 2]
```diff
@@ -1219,8 +1221,8 @@
  */
 'excluded_directories' =>
 	array (
-		'.snapshot',
-		'~snapshot',
+		'^\.snapshot$',
```
Comment (mmattel) [inline config/config.sample.php]: * I propose that you leave the original examples as written and add additional but different more fancy once to the list (blacklisted and excluded). Helps to make things more clear that you can use the old and the new (regex) way.
* In addition you must add a descriptive text that regex is allowed.
Add a `WARNING: ` message about the possible performance impact like:
```
WARNING: COMPLEX REGEX MAY SIGNIFICANT IMPACT YOUR SCANNING PERFORMANCE
```
The term warning + the colon + blank afterwards must be exacly written as in the example above because this has a doc rendering background.

File: lib/private/Files/Filesystem.php [hunk 3]
```diff
@@ -655,20 +655,26 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
...
+                                foreach($excluded as $blackitem) {
+                                        if(preg_match('/'.$blackitem.'/i', $path_part)) {
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Please use: `preg_match` --> `\preg_match` (and below)

See: https://doc.owncloud.com/server/10.3/developer_manual/core/code-standard.html

**Comment by mmattel** [comment]
Your `Fixes` list from above is not correct as sync excludes are not addressed by this PR.
These are a different topic. Pls remove them

File: lib/private/Files/Filesystem.php (hunk 3 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Thanks for the doc's reference, correction is done.

File: config/config.sample.php (hunk 2 above)
Comment (tdaget) [reply config/config.sample.php]: Can't leave previous examples as they must be regex compliant (per example the dot in '.snapshot'). But more information and example might be welcome.
Warning added. Thanks

File: lib/private/Files/Filesystem.php (hunk 1 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Advise 2 and 3: done.
Advise 1: don't know how to implement them ...

**Comment by tdaget** [comment]
> Your `Fixes` list from above is not correct as sync excludes are not addressed by this PR.
> These are a different topic. Pls remove them

Done

File: lib/private/Files/Filesystem.php [hunk 4]
```diff
@@ -658,16 +658,30 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
 		if ($excluded) {
 			$excluded = \array_map('trim', $excluded);
 			$excluded = \array_map('strtolower', $excluded);
-			$match = \array_intersect($path_parts, $excluded);
-			if ($match) {
-				return true;
+			foreach($excluded as $blackitem) {
+				if(@\preg_match($blackitem, null) === false) {
+					\OC::$server->getLogger()->error('Malformed exclude regex: '.$blackitem.' - Check excluded_directories variable in config file.', ['app' => __CLASS__]);
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: 1. Great that there is an error message now.
The message itelf only shows that something bad happened but gives no clue what it was.
It would be good to **add** a better hint about the cause. Please see the example below:
Based on:
https://www.php.net/manual/de/function.preg-last-error.php and https://stackoverflow.com/questions/4440626/how-can-i-validate-regex/12941133#12941133
I allowed myself to add/rename some stuff...
```
function is_preg_error()
{
    $errors = array(
        PREG_NO_ERROR               => 'Code 0 : No errors',
        PREG_INTERNAL_ERROR         => 'Code 1 : Internal PCRE error, most likely a syntax error',
        PREG_BACKTRACK_LIMIT_ERROR  => 'Code 2 : Backtrack limit was exhausted',
        PREG_RECURSION_LIMIT_ERROR  => 'Code 3 : Recursion limit was exhausted',
        PREG_BAD_UTF8_ERROR         => 'Code 4 : The offset did not correspond to the begin of a valid UTF-8 code point',
        PREG_BAD_UTF8_OFFSET_ERROR  => 'Code 5 : Malformed UTF-8 data',
        PREG_JIT_STACKLIMIT_ERROR   => 'Code 6 : Just-in-time compiler stack limit reached',
    );

    return $errors[\preg_last_error()];
}
```

File: lib/private/Files/Filesystem.php [hunk 5]
```diff
@@ -658,16 +658,30 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
...
+				} else {
+					foreach($path_parts as $path_part) {
+						if(\preg_match('/'.$blackitem.'/i', $path_part)) {
+							return true;
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: You may want to add a check if regex processing was ok:
`\preg_last_error() !== PREG_NO_ERROR`
and log an error in case using the function I described above.

**Comment by mmattel** [comment]
1. There are still coding style errors, pls see https://doc.owncloud.com/server/10.3/developer_manual/core/code-standard.html
and the build failing details from drone above

2. I am still not convinced hard switching from `array_intersect` done in one shot to `preg_match` double looping by default. I highly recommend as written in my comments above to at least do some checking for common regex characters and decide based on the result the way of checking. I know that there is no 100% identifyer, but I refrenced a hint in my comment. Things are easier for Windows backends as they have more excluding characters in filenames. But this is a matter of a proper description. This was the reason for keeping the old examples and adding new regex once.

File: lib/private/Files/Filesystem.php (hunk 5 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Done

File: lib/private/Files/Filesystem.php (hunk 4 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Your code is added.
I put it directly in Filesystem.php class, don't know where to put it in a utility class ...

**Comment by tdaget** [comment]
@mmattel 
I agree, switching from `array_intersect` to `preg_match` might induce high CPU consumption.
I think we can consider following solutions:
1- create a new `excluded_directories` and a new `blacklisted_files` dedicated to regex match.
    I am scared of configuration difficulties and misunderstanding by users, I prefer to avoid this situation
2- allow a mix of strict filename and regex in these configuration variables, for each test is this is a string, if yes search for a direct match, if not try regex match. Double work ... And I don't know how to identify a simple string and invalid regex
3- consider that strict filename matching (previous way of working) is cover by regex use with an acceptable CPU cost
    The way I'm currently considering. Plus it permits to check regex validity systematically.

You propose to search for common regex characters, maybe not so reliable as these characters (for many of them and not least) can be used in filenames.

But all advice is welcome :-/

**Comment by tdaget** [comment]
An idea to speed up the regex check: no need to check all `blacklisted_files` to all part of 'path to file' array, we need to match only the filename part.
Question (as I'm not an expert PHP developer): in the path_parts array variable, is the last item is ALWAYS (the always is very important here) the filename itself? In that case, we can update code with -drop foreach line 709, and replace `if (\preg_match('/'.$blackitem.'/i', $path_part)) {` by `if (\preg_match('/'.$blackitem.'/i', end($path_parts))) {` line 710.

**Comment by mmattel** [comment]
Hope you are fine with all my comments... :smile:
1. You need to rebase as there was a change in the drone setup that needs to be populated into your PR so drone can process properly and turns into green 
2. `foreach ($excluded as $blackitem)` and `foreach ($blacklist as $blackitem)` :
only an optical thing, but to avoid confusion, please use a more neutral name for `$blackitem`
3. The code for checking `\preg_last_error()` should also run after each `foreach ($path_parts`.
This because you can run in all the other errors that syntax which is checked before. Recommending that the check and the error is put into a function which is called with an (optional) error text as you have written it already. This would reduce number of lines and makes the code more readble.
4. I get your point regarding regex / non regex confusion. To keep the positives of both variants, what about the idea to implement two new config parameters like `blacklisted_files_regex` and `excluded_directories_regex`? You then can run the code based on which one is set. Yes, you can set both but it gives admins the ability deciding which way is best for them. If both are set, you must run both and combine the result with `or`. Please a good descriptive text if you go for that approach.
5. Thanks for looking at `integrity.excluded.files`, but pls turn that back, as there is another PR already in review which takes care on that (and more).
6. Regarding `blacklisted_files`:
-Reminder, this code **must** always run, but not necessarily a empty regex version... (see idea above)
-The last element of `$path_parts` can either be the directoryname or the filename if a filename exists. You can improve the code as you thought and do the same for the current way if you want. Thanks for spotting that.

**Review by tdaget — COMMENTED** [review]
Typo corrected in tests/data/app/invalid-info.xml
Add regular match for blacklisted folders and blacklisted files

File: apps/federatedfilesharing/templates/settings-personal-general.php [hunk 6]
```diff
@@ -44,7 +44,7 @@ class='js-gs-share social-gnu'>
 				Facebook
 			</button>
 			<button class="social-googleplus pop-up"
-				data-url='https://plus.google.com/share?url=<?php p(\urlencode($_['reference'])); ?>'/>
+				data-url='https://plus.google.com/share?url=<?php p(\urlencode($_['reference'])); ?>'>
```
Comment (mmattel) [inline apps/federatedfilesharing/templates/settings-personal-general.php]: Great for finding bugs / creating improvements.
As this one has nothing to do with regex, it must be put into a own PR.

File: config/config.sample.php [hunk 7]
```diff
@@ -1201,32 +1201,73 @@
...
+ */
+'blacklisted_files_regex' => array (
+//	'.*\.pst$',		// disallow *.pst (MS Outlook) file
+//	'.*dummy.*',	// disallow file with name contaning the string 'dummy'
+//	'^sample.*',	// disallow file with name beginning by 'sample'
+),
+
+
+/**
+ * Blacklist specific directory names and disallow scanning, creating and renaming
```
Comment (mmattel) [inline config/config.sample.php]: `Blacklist specific directory` --> `Exclude specific directory`

File: config/config.sample.php [hunk 8]
```diff
@@ -1201,32 +1201,73 @@
...
  */
-'excluded_directories' =>
-	array (
+'excluded_directories' => array (
+	'.snapshot',	// exclude '.snapshot' folder
+	'~snapshot',	// exclude '~snapshot' folder
+),
+
+/**
+ * Blacklist specific directory names based on regex expression. Scanning, creating and renaming
```
Comment (mmattel) [inline config/config.sample.php]: `Blacklist specific directory` --> `Exclude specific directory`

File: config/config.sample.php [hunk 9]
```diff
@@ -1201,32 +1201,73 @@
...
+ * WARNING: COMPLEX REGEX MAY SIGNIFICANT IMPACT YOUR SCANNING PERFORMANCE
+ * WARNING: String case is systematicaly lowered
+ */
+'excluded_directories_regex' => array (
+//	'.*backup.*',	// exclude folder with name containing the string 'backup'
+//	'Thomas.*',		// exclude folder with name beginning by 'Thomas'
+),
+
+/**
+ * Exclude specific directory names and disallow scanning, creating and renaming
```
Comment (mmattel) [inline config/config.sample.php]: This block seems to be a leftover and forgotten to be deleted...

File: config/config.sample.php [hunk 10]
```diff
@@ -1201,32 +1201,73 @@
...
-	),
+),
+
+
 /**
  * Exclude files from the integrity checker command
  */
 'integrity.excluded.files' =>
-	array (
+	array(
```
Comment (mmattel) [inline config/config.sample.php]: As already stated, pls no change here as we already have an open PR which covers this issue (plus other stuff).

File: lib/private/Files/Filesystem.php [hunk 11]
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
 	}
 
 	/**
-	* Check if the directory path / file name contains a Blacklisted or Excluded name
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Pls keep this explanation text from the content point of view as it describes what is meant with each kind.
It is totally ok to move it down to line 688.

File: settings/templates/help.php [hunk 12]
```diff
@@ -45,11 +45,12 @@
 	<?php
 	} ?>
 
-	<li>
-		<a href="https://owncloud.com/subscriptions/" target="_blank" rel="noreferrer">
-			<?php p($l->t('Commercial support')); ?> ↗
-		</a>
-	</li>
+		<li>
```
Comment (mmattel) [inline settings/templates/help.php]: As stated above, this has nothing to do with regex.
Pls create an own PR for fixing other issues than regex.

File: tests/data/app/invalid-info.xml [hunk 13]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
-<info
+<info>
```
Comment (mmattel) [inline tests/data/app/invalid-info.xml]: As stated above, this has nothing to do with regex.
Pls create an own PR for fixing other issues than regex.

File: lib/private/Files/Filesystem.php [hunk 14]
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
...
+								.self::is_preg_error(),
+								['app' => __CLASS__]);
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Manage regex blacklist against Folder of File.
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Remove the word `blacklist`

File: lib/private/Files/Filesystem.php [hunk 15]
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
...
+	 * @param blacklist Blacklist regex array as defined in Config file
+	 * @param path Folder of File complete path contained in an array
+	 * @return boolean folder of file forbudden status
+	 */
+	private static function blacklistRegexAgainstFolderOrFile($blacklist, $path) {
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Please use a neutral name instead of `$blacklist`

File: lib/private/Files/Filesystem.php [hunk 16]
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
...
+		}
+
+		return true;
+	}
+
+	/**
+	 * Manage regex blacklist against Folder of File.
+	 * @param blacklist Blacklist regex array as defined in Config file
+	 * @param path Folder of File complete path contained in an array
+	 * @return boolean folder of file forbudden status
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: `folder of file forbudden status` --> `folder of file regex status`

**Comment by mmattel** [comment]
We are going into the right direction :smile:
I see that you are a big fan of the term blacklist...
Using it as a common name everywhere is a bad idea. The original naming was made by design.
- Have function and variable names based on the usecase which further reflects config.sample.
(means, use blacklist... where you run code for blacklist and excluded... where you run code for excluded)
- `blacklistRegexAgainstFolderOrFile`--> `checkRegexAgainstFolderOrFile`
- `regex_validy_check($regexToBeChecked)` --> `regex_validy_check($regexToBeChecked, $message="")`
Add a second parameter and handover the message via the caller (like `Check excluded_directories variable in config file`). This message differs either from blacklist or excluded. Rename the first part of the message `Exclude regex error:` --> `Regex error:`
- Based on the example `blacklist_folders` and `blacklist_folders_regex` (same for excluded), as stated above, you **must** run them if set and combine the results with `or` for a final return.

File: config/config.sample.php (hunk 7 above)
Comment (tdaget) [reply config/config.sample.php]: Corrected

File: config/config.sample.php (hunk 8 above)
Comment (tdaget) [reply config/config.sample.php]: Corrected

File: config/config.sample.php (hunk 9 above)
Comment (tdaget) [reply config/config.sample.php]: You're alright
Corrected

File: config/config.sample.php (hunk 10 above)
Comment (tdaget) [reply config/config.sample.php]: Can't see this point in my code ... 'll check after correcting all other points

File: lib/private/Files/Filesystem.php (hunk 14 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Word dropped

File: lib/private/Files/Filesystem.php (hunk 15 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Corrected

File: lib/private/Files/Filesystem.php (hunk 16 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Changed

**Comment by tdaget** [comment]
> We are going into the right direction 
> I see that you are a big fan of the term blacklist...
> Using it as a common name everywhere is a bad idea. The original naming was made by design.
> 
> * Have function and variable names based on the usecase which further reflects config.sample.
>   (means, use blacklist... where you run code for blacklist and excluded... where you run code for excluded)
> * `blacklistRegexAgainstFolderOrFile`--> `checkRegexAgainstFolderOrFile`
Changed

> * `regex_validy_check($regexToBeChecked)` --> `regex_validy_check($regexToBeChecked, $message="")`
>   Add a second parameter and handover the message via the caller (like `Check excluded_directories variable in config file`). This message differs either from blacklist or excluded. Rename the first part of the message `Exclude regex error:` --> `Regex error:`
Good idea
Corrected

> * Based on the example `blacklist_folders` and `blacklist_folders_regex` (same for excluded), as stated above, you **must** run them if set and combine the results with `or` for a final return.
Sorry, Can't understand this point.
As I modified the isForbiddenFileOrDir method the sequence of tests (exclude then exclude_regex then blacklist then blacklist_regex) do by construction an implicit OR ==> first test make a 'return true' and kill all other tests. And each test is done only if directive is set in config file.
In my understanding this permit to skip all non-mandatory work.
Why do you want to combine tests ? I'll make code less readable ...

> * Please rebase to make CI happy and green
I'm working on it, step after step, but lessons are hard to remember ...

File: apps/federatedfilesharing/templates/settings-personal-general.php (hunk 6 above)
Comment (tdaget) [reply apps/federatedfilesharing/templates/settings-personal-general.php]: Reverted

File: settings/templates/help.php (hunk 12 above)
Comment (tdaget) [reply settings/templates/help.php]: Reverted

File: tests/data/app/invalid-info.xml (hunk 13 above)
Comment (tdaget) [reply tests/data/app/invalid-info.xml]: Reverted

File: lib/private/Files/Filesystem.php [hunk 17]
```diff
@@ -625,28 +625,92 @@ public static function isIgnoredDir($dir) {
...
+				}
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	* Check if the directory / file name is blacklisted in config.php.
+	* $ed and the query with can be redesigned if filesystem.php will get
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Please use at least parts the original description (written in one of my comments already)
I have made a suggestion and added some words to blacklist to extend the usecase description
```
	/**
	* Check if the directory path / file name contains a Blacklisted or Excluded name
	* config.php parameter arrays can contain file names to be blacklisted or directory names to be excluded
	* Blacklist ... files that may harm the owncloud environment like a foreign .htaccess file or other unwanted files
	* Excluded  ... directories that are excluded from beeing further processed, like snapshot directories
	* $ed and the query with can be redesigned if filesystem.php will get a constructor where it is then possible 
	* to define the excluded directory names for unit tests
```

File: lib/private/Files/Filesystem.php [hunk 18]
```diff
@@ -655,20 +719,32 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
...
+			}
+		}
+		if ($blacklist_files) {
+			if (\array_intersect($blacklist_files, \end($path_parts))) {
+				return true;
+			}
 		}
+		if ($blacklist_files_regex) {
+			if (self::checkRegexAgainstFolderOrFile($blacklist_files_regex, \end($path_parts),
+													"Check blacklisted_files_regex variable in config file")) {
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: The intend is a bit too big :smile:

File: config/config.sample.php [hunk 19]
```diff
@@ -1217,23 +1234,38 @@
...
-		'~snapshot',
-	),
+'excluded_directories' => array (
+	'.snapshot',	// exclude '.snapshot' folder
+	'~snapshot',	// exclude '~snapshot' folder
+),
+
+/**
+ * Exclude specific directory names based on regex expression. Scanning, creating and renaming
+ * is disabled when folder match one of the regex.
```
Comment (mmattel) [inline config/config.sample.php]: Just put this into two lines for a better readbility:
```
 * Exclude specific directory names based on regex expression.
 * Scanning, creating and renaming is disabled when folder matches one of the regex.
```

File: config/config.sample.php [hunk 20]
```diff
@@ -1217,23 +1234,38 @@
...
+ * Use of excluded_directories variable have precedence (evaluated before) on excluded_directories_regex variable.
```
Comment (mmattel) [inline config/config.sample.php]: `Use of...` --> `INFO: Use of ...`
The term info or warnig has a special meaning here.
It will be rendered in the documentation in a special optical way.

File: config/config.sample.php (hunk 10 above)
Comment (mmattel) [reply config/config.sample.php]: pls re-check. still there...

**Comment by mmattel** [comment]
Very good! only a view things to take care on :smile:

File: lib/private/Files/Filesystem.php [hunk 21]
```diff
@@ -625,28 +625,92 @@ public static function isIgnoredDir($dir) {
...
+	 * @return boolean folder of file regex status
+	 */
+	private static function checkRegexAgainstFolderOrFile($regexList, $path, $message="") {
+		foreach ($regexList as $item) {                               // foreach given regex
+			\preg_match($item, null);                                 // regex validy check
+			if (self::regexValidyCheck($item, $message)) {            // check if regex error occur
+
+				foreach ($path as $path_part) {                       // foreach folder or file in given path
+					if (\preg_match('/'.$item.'/i', $path_part)       // regex match item
+							&& self::regexValidyCheck($item)) {       // check if regex error occur
```
[the remark this answers is not in the document]
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Good idea, added.

File: lib/private/Files/Filesystem.php (hunk 17 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Sorry, dropped this block by mistake
Corrected

File: lib/private/Files/Filesystem.php [hunk 22]
```diff
@@ -655,20 +719,32 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
...
 			}
 		}
-		$blacklist = \array_map('trim', $blacklist);
-		$blacklist = \array_map('strtolower', $blacklist);
-		$match = \array_intersect($path_parts, $blacklist);
-		if ($match) {
-			return true;
+		if ($exclude_folders_regex) {
+			if (self::checkRegexAgainstFolderOrFile($exclude_folders_regex, $path_parts,
+													"Check excluded_directories_regex variable in config file")) {
```
[the remark this answers is not in the document]
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Corrected

File: lib/private/Files/Filesystem.php (hunk 18 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Corrected

File: config/config.sample.php (hunk 19 above)
Comment (tdaget) [reply config/config.sample.php]: Corrected

File: config/config.sample.php (hunk 20 above)
Comment (tdaget) [reply config/config.sample.php]: Added

File: config/config.sample.php [hunk 23]
```diff
@@ -1201,12 +1201,29 @@
...
+ */
+'blacklisted_files' => array (
+	'.htaccess',	// disallow .htaccess file
+),
+
 /**
- * Blacklist a specific file or files and disallow the upload of files
- * with this name. `.htaccess` is blocked by default.
+ * Blacklist filenames based on regex expression.
+ * Use of blacklisted_files variable have precedence (evaluated before) on blacklisted_files_regex variable.
```
[the remark this answers is not in the document]
Comment (tdaget) [reply config/config.sample.php]: Added

File: config/config.sample.php [hunk 24]
```diff
@@ -1201,12 +1201,29 @@
...
+ * WARNING: USE THIS ONLY IF YOU KNOW WHAT YOU ARE DOING.
+ */
+'blacklisted_files' => array (
+	'.htaccess',	// disallow .htaccess file
+),
+
 /**
- * Blacklist a specific file or files and disallow the upload of files
- * with this name. `.htaccess` is blocked by default.
+ * Blacklist filenames based on regex expression.
```
[the remark this answers is not in the document]
Comment (tdaget) [reply config/config.sample.php]: Corrected

**Comment by mmattel** [comment]
- `integrity.excluded.files` seems still not reverted, maybe just overseen.
- in config.sample, in the new regex examples, no need to put the array elements as remarks (ok for the description). 
- maybe a good idea to squash the commits now

File: lib/private/Files/Filesystem.php [hunk 25]
```diff
@@ -625,28 +625,102 @@ public static function isIgnoredDir($dir) {
...
+	}
+
+	/**
+	 * Regex validy check.
+	 * @param string $regexToBeChecked regex to be checked
+	 * @param string $message message given in log in case of invalid regex
+	 * @return boolean
+	 */
+	private static function regexValidyCheck($regexToBeChecked, $message="") {
+		if (self::is_preg_error() !== PREG_NO_ERROR) {
```
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Shouldnt this be `\preg_last_error()` as `is_preg_error` returns a string ?

**Comment by tdaget** [comment]
> * in config.sample, in the new regex examples, no need to put the array elements as remarks (ok for the description).

No need? Maybe but as sysadmin I know examples are always welcome, especially with regex

If you want absolutely to remove these lines of example maybe we can put them in comment block?

File: lib/private/Files/Filesystem.php (hunk 25 above)
Comment (tdaget) [reply lib/private/Files/Filesystem.php]: Right

**Comment by tdaget** [comment]
> * maybe a good idea to squash the commits now

Need some help for this (don't understand the "squash commits" ??)

**Comment by mmattel** [comment]
I meant: `//    '.*\.pst$',     // di...` —> `    '.*\.pst$',     // di...`
It is removing the first two `//`

**Comment by mmattel** [comment]
Example squashing the last three commits:
(https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git)
```
git reset --soft HEAD~3 &&
git commit
git push -f
````
At the moment you have 32 commits... (just see the number on top).
If you squash all 32 commits and push them, you see finally only one.
Note: the number of commits to squash must not exceed the existing number of commits of this PR. You do squashing when the work is close to final / or final already. This makes checking easier.

**Comment by mmattel** [comment]
on top, pls set the checkmark that you have added unit tests

**Comment by tdaget** [comment]
> on top, pls set the checkmark that you have added unit tests

Done

**Comment by tdaget** [comment]
> I meant: `// '.*\.pst$', // di...` —> ` '.*\.pst$', // di...`
> It is removing the first two `//`

Remove the two `//` mean activating this regex by default (as everybody bases his config.php on the sample file). I'm afraid of confusion and questions for a lot of people.

At most I can consider doing this change if the `*.pst` regex as Outlook files raises a lot of undesirable network traffic and Owncloud activities.

**Comment by mmattel** [comment]
> activating this regex by default

No worry about that one. The name says **sample** and there is a description in the documentation not to just copying stuff but using it wisely... And only a limited set config options are used during installation. Anything else must be used by actively adding/configuring it to the running config file.

**Comment by mmattel** [comment]
I just tried to help a bit and wanted to clone your branch but have seen that you did not used a branch.
Now this is somehow a bad thing because you will hardly come into sync with master and its ongoing changes without having a branch. But this is needed because of changes running CI. My proposal therefore would be the following:
- Create a branch from actual master and make/copy all the changes you did in this PR.
- Push it and make a new PR from that push.
- Close this PR

Big benefit: Clean, short and open for changes in master (easy rebaseable) and CI will turn green
If you cant do that, I am willing to help, but can do this earliest late evening today. Just let me know.

Great work!

**Comment by mmattel** [comment]
Closing in favour of https://github.com/owncloud/core/pull/36360.
I have improved your changes. Tested and runs like a charm :smile:
@tdaget please do any changes if necessary in #36360

**Comment by tdaget** [comment]
> Closing in favour of #36360.
> I have improved your changes. Tested and runs like a charm 
> @tdaget please do any changes if necessary in #36360

Thanks for doing this.
I must admit that I was a little lost in the management of branches and PRs.
I'm still not comfortable with git / github ...

**Comment by mmattel** [comment]
You could use smartgit (https://www.syntevo.com/smartgit/download/) available for various OS.
Makes life a lot easier when you start using git

**Comment by tdaget** [comment]
Thanks for our advise.
I start to use github with the official 'github desktop' under linux
not so bad once we understand the basic principles ...

Thanks for your help in this PR
I've close all my branch/PR created in a desperate discover of tools. I created another branch named `Typo-corrections` in order to work on all small typo found during this PR.
